stdlib/public/Cxx/CxxConvertibleToCollection.swift

```diff
@@ -0,0 +1,70 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// A C++ type that can be converted to a Swift collection.
+public protocol CxxConvertibleToCollection {
+  associatedtype RawIterator: UnsafeCxxInputIterator
+
+  /// Do not implement this function manually in Swift.
+  mutating func __beginUnsafe() -> RawIterator
+
+  /// Do not implement this function manually in Swift.
+  mutating func __endUnsafe() -> RawIterator
+}
+
+@inlinable @inline(__always)
+internal func forEachElement<C: CxxConvertibleToCollection>(
+  of c: C,
+  body: (C.RawIterator.Pointee) -> Void
+) {
+  var mutableC = c
+  withExtendedLifetime(mutableC) {
+    var rawIterator = mutableC.__beginUnsafe()
+    let endIterator = mutableC.__endUnsafe()
+    while rawIterator != endIterator {
+      body(rawIterator.pointee)
+      rawIterator = rawIterator.successor()
+    }
+  }
+}
+
+extension Array {
```

Comment:
It would be really cool if we could provide optimized implementations for contiguous collections (C++ provides this information in iterator traits). This could just be a reserve + memcpy. I don't know if LLVM is smart enough to do that on its own (probably not).

That can be a follow up patch, though.

Reply:
Great point! I'll do that in a separate patch.